codec/src/main/java/io/netty5/handler/codec/MessageToByteEncoderForBuffer.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * <p>Example implementation which encodes {@link Integer}s to a {@link Buffer}.
+ *
+ * <pre>
+ *     public class IntegerEncoder extends {@link MessageToByteEncoderForBuffer}&lt;{@link Integer}&gt; {
+ *         {@code @Override}
+ *         public void encode({@link ChannelHandlerContext} ctx, {@link Integer} msg, {@link Buffer} out)
+ *                 throws {@link Exception} {
+ *             out.writeInt(msg);
+ *         }
+ *     }
+ * </pre>
+ */
+public abstract class MessageToByteEncoderForBuffer<I> extends ChannelHandlerAdapter {
+
+    private final TypeParameterMatcher matcher;
+    private final boolean preferDirect;
+
+    /**
+     * see {@link #MessageToByteEncoderForBuffer(boolean)} with {@code true} as boolean parameter.
+     */
+    protected MessageToByteEncoderForBuffer() {
+        this(true);
+    }
+
+    /**
+     * see {@link #MessageToByteEncoderForBuffer(Class, boolean)} with {@code true} as boolean value.
+     */
+    protected MessageToByteEncoderForBuffer(Class<? extends I> outboundMessageType) {
+        this(outboundMessageType, true);
+    }
+
+    /**
+     * Create a new instance which will try to detect the types to match out of the type parameter of the class.
+     *
+     * @param preferDirect {@code true} if a direct {@link Buffer} should be tried to be used as target for
+     *                     the encoded messages. If {@code false} is used it will allocate a heap {@link Buffer}.
+     */
+    protected MessageToByteEncoderForBuffer(boolean preferDirect) {
+        matcher = TypeParameterMatcher.find(this, MessageToByteEncoderForBuffer.class, "I");
+        this.preferDirect = preferDirect;
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * @param outboundMessageType The type of messages to match.
+     * @param preferDirect        {@code true} if a direct {@link Buffer} should be tried to be used as target for
+     *                            the encoded messages. If {@code false} is used it will allocate a heap {@link Buffer}.
+     */
+    protected MessageToByteEncoderForBuffer(Class<? extends I> outboundMessageType, boolean preferDirect) {
+        matcher = TypeParameterMatcher.get(outboundMessageType);
+        this.preferDirect = preferDirect;
+    }
+
+    /**
+     * Returns {@code true} if the given message should be handled. If {@code false} it will be passed to the next
+     * {@link ChannelHandler} in the {@link ChannelPipeline}.
+     */
+    public boolean acceptOutboundMessage(Object msg) throws Exception {
+        return matcher.match(msg);
+    }
+
+    @Override
+    public Future<Void> write(ChannelHandlerContext ctx, Object msg) {
+        Buffer buf = null;
+        try {
+            if (acceptOutboundMessage(msg)) {
+                @SuppressWarnings("unchecked")
+                I cast = (I) msg;
+                buf = allocateBuffer(ctx, cast, preferDirect);
```

Comment:
Another possibility, as opposed to allocating for each message, could be to allocate larger buffers, keep the buffer around in a field, and shave off the encoded part with `split()`. Then `readableBytes() == 0` also wouldn't be a special case.

Perhaps it depends a bit on the use case which approach is best? Something to think about at least.